### PR TITLE
refactor: refactor simple indexer

### PR DESCRIPTION
**Pull request by winstonww**
Opening this draft request to address the changes as discussed during the pairing session with @Yongxuanzhang and @hanxiao. One more unit test to test the `update` method needs to be added. PR should be ready for review once the remaining test is added. Thank you.

### CODE REVIEW

**Comment by winstonww** [comment]
@Yongxuanzhang @hanxiao 

So a potential issue is observed after we remove `traversal_paths` option from `SimpleIndexer`.

Consider the case where we want to search and rank some documents with chunks. Using the original `SimpleIndexer` with `traversal_paths` option enabled, we can pass these docs to it and the output will be docs with the same hierarchy/structure, but with each chunk of each doc having `matches` appended to it. We then pass the output docs of `SimpleIndexer` to a ranker (e.g. `SimpleRanker`) so we get the matches of each top-level `doc`.

Now after `traversal_paths` option removed from `SimpleIndexer`, we first pass the docs to something like a `TraversalExecutor` to traverse and flatten the docs so that the new `SimpleIndexer` gets the input with the expected structure (flattened chunk docs) since we need to first match at the chunk level. The output of the `SimpleIndexer` will be flattened chunk docs with matches. Now, the question is, how do we rank the top level query docs given that docs are already flattened at the chunk level and the hierarchy/structure is lost? We cannot simply pass the flatten chunk docs to the ranker because what ranker needs to return is the top-level docs with matches. 

Feedback on this is appreciated, thank you.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+# install requirements before copying the workspace
+COPY requirements.txt /requirements.txt
+RUN pip install --default-timeout=1000 --compile -r requirements.txt
+
+# setup the workspace
+COPY . /workspace
+WORKDIR /workspace
+
+ENTRYPOINT ["jina", "executor", "--uses", "config.yml"]
```
Comment (nan-wang) [inline Dockerfile:11]: We don't need `Dockerfile` in this case. The hubble will take care of it in this simple case.

File: executor.py [hunk 2]
```diff
@@ -13,12 +13,13 @@ class SimpleIndexer(Executor):
...
-            self,
-            match_args: Optional[Dict] = None,
-            **kwargs,
+        self,
+        match_args: Optional[Dict] = None,
+        **kwargs,
     ):
         """
         Initializer function for the simple indexer
+        :param match_args: the arguments to `Document`'s match function
```
Comment (nan-wang) [inline executor.py:22]: This line should go into the docstrings of the class

**Review by nan-wang — CHANGES_REQUESTED** [review]

**Comment by nan-wang** [comment]
> how do we rank the top level query docs given that docs are already flattened at the chunk level and the hierarchy/structure is lost?

After flattening, the `granularity` information is preserved. Do you have some MWE codes to show your concerns?

File: Dockerfile (hunk 1 above)
Comment (winstonww) [reply Dockerfile]: Addressed, please see [here](https://github.com/jina-ai/executor-simpleindexer/pull/1/commits/6f12e89a4194f5cb85751ba0deca9f7443d97274) for the change.

File: executor.py (hunk 2 above)
Comment (winstonww) [reply executor.py]: Moved, please see [here](https://github.com/jina-ai/executor-simpleindexer/pull/1/commits/6f12e89a4194f5cb85751ba0deca9f7443d97274), thank you.

**Comment by winstonww** [comment]
Yes, `granularity` is preserved, however we no longer have access to the top level docs after we traversed the docs. Consider the following MWE. In the original case:

```
import numpy as np
from jina import Document, DocumentArray, Flow 

doc1 = Document(id='doc1')
doc1.chunks.append(Document(id='doc1-chunk1', embedding=np.array([1,0,0,0])))
doc1.chunks.append(Document(id='doc1-chunk2', embedding=np.array([0,1,0,0])))

doc2 = Document(id='doc2')
doc2.chunks.append(Document(id='doc2-chunk1', embedding=np.array([0,0,1,0])))
doc2.chunks.append(Document(id='doc2-chunk2', embedding=np.array([0,0,0,1])))

docs = DocumentArray([doc1, doc2])

# ORIGINAL SCENARIO: Using the existing SimpleIndexer with `default_traversal_paths`
with (
        Flow()
        .add(uses='jinahub://SimpleIndexer',uses_with={'default_traversal_paths': ['c']})
        .add(uses='jinahub://SimpleRanker')
    ) as f:
    f.index(docs)
    result_docs = f.search(docs, return_results=True)[0].data.docs
    assert result_docs[0].id == 'doc1' and result_docs[0].matches[0].id == 'doc1'
    assert result_docs[1].id == 'doc2' and result_docs[1].matches[0].id == 'doc2'
```
In the above case, ranker can still have access to top level docs, hence it can successfully match and rank `doc1` and `doc2`. In the new scenario where `traversal_paths` option is removed from `SimpleIndexer`:


```
from executor import SimpleIndexer

class TraverseExecutor(Executor):
    def __init__(self, default_traversal_paths, **kwargs):
        super().__init__(**kwargs)
        self.default_traversal_paths = default_traversal_paths

    @requests
    def traverse(self, docs, **kwargs):
        return docs.traverse_flat(self.default_traversal_paths)

class Reconstructor(Executor):
    @requests(on='/search')
    def reconstruct(self, docs, **kwargs):
        # since `docs` are chunk level docs, how do we get the parent docs?
        pass

# NEW SCENARIO: With traversal_paths option removed
with (
        Flow()
        .add(uses=TraverseExecutor, uses_with={'default_traversal_paths': ['c']})
        .add(uses=SimpleIndexer)
        .add(uses=Reconstructor)
        .add('jinahub://SimpleRanker')
    ) as f:
    f.index(docs)
    result_docs = f.search(docs, return_results=True)[0].data.docs
    # result_docs do not have 'doc1' and 'doc2', but only 'doc1-chunk1', 'doc1-chunk2', ... etc
    # because docs are traversed and flattened at the chunk lavel after `TraverseExecutor` processed the docs

```
I am not sure how we can implement the `Reconstructor` to recover the top-level docs since the input docs are flattened chunk docs. And if the `Reconstructor` cannot be implemented, then ranker in this case does not have access to top-level docs (i.e. `doc1` and `doc2`) and won't be able to match and rank `doc1` and `doc2`.

File: tests/test_simple_indexer.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    # test search from empty indexed docs
+    shutil.rmtree(tmpdir)
+    indexer = SimpleIndexer(metas=metas)
+    indexer.index(DocumentArray())
+    indexer.search(docs)
+    for doc in docs:
+        assert not doc.matches
+
+    # test search empty docs
+    indexer.search(DocumentArray())
```
Comment (hanxiao) [inline tests/test_simple_indexer.py:170]: remove this test

File: README.md [hunk 4]
```diff
@@ -1 +1,23 @@
...
+#### via source code
+
+```python
+from jina import Flow
+	
+f = Flow().add(uses='jinahub://SimpleIndexer')
+```
+
+- To override `__init__` args & kwargs, use `.add(..., uses_with: {'key': 'value'})`
+- To override class metas, use `.add(..., uses_metas: {'key': 'value})`
```
Comment (hanxiao) [inline README.md:23]: unnecessary, delete it
```suggestion
```

Comment (hanxiao) [reply README.md:23]: e.g. SimpleIndexer will store the documents in a folder that is specified by `workspace` (which can be defined via ...). Inside this `workspace` there are two files: `head.bin`, `body.bin` which corresponds to the file structure of `DocumentArrayMemmap`

Comment (hanxiao) [reply README.md:23]: two designs of SimpleIndexer
1. it leverages the `DAM` in the core
2. it leverages the `.match` in the core


To leverage the full flexibility of `.match`, one can specify `match_args` in `SimpleIndexer`, e.g. 

```python

f = Flow().add(uses=SimpleIndexer, uses_with={'match_args': {'lhs_traversal_path': ['c', 'm']}})

```

File: tests/test_simple_indexer.py (hunk 3 above)
Comment (winstonww) [reply tests/test_simple_indexer.py]: Addressed, please see [here](https://github.com/jina-ai/executor-simpleindexer/pull/1/commits/3aee3434b2a5fae552a35014e6fb32509f9f9c03) for the change.

File: config.yml [hunk 5]
```diff
@@ -4,3 +4,4 @@ with:
 metas:
   py_modules:
     - executor.py
+  workspace: workspace/
```
Comment (hanxiao) [inline config.yml:7]: i dont think this is necessary
```suggestion
```

Comment (winstonww) [reply config.yml:7]: Hmm, If we remove the above, running `jina executor --uses config.yml` will result in:
```ValueError: Neither `metas.workspace` nor `runtime_args.workspace` is set, are you using this Executor is a Flow?```

Comment (hanxiao) [reply config.yml:7]: keep it in this way then, i think it is a bug in the core

Comment (winstonww) [reply config.yml:7]: okie, I will merge this PR if there are no other issues

**Pull request merged**